software/contrib/pams.py

```diff
@@ -0,0 +1,1118 @@
+                    self.e_position = self.e_position + 1
+                    if self.e_position >= len(self.e_pattern):
+                        self.e_position = 0
+
+        self.cv_out.voltage(out_volts)
+
+class SettingChooser:
+    """Menu UI element for displaying a Setting object and the options associated with it
+    """
+    def __init__(self, prefix, setting, gfx=None, submenu=[]):
+        """Create a setting chooser for a given item
+
+        @param prefix  A prefix we show before the option title (e.g. 'CV1 | ')
+        @param setting  The Setting object we're manipulating
+        @param submenu  A list of SettingChooser items that make up this setting's submenu
+        @param gfx  A list of 12x12 pixel bitmaps we can optionally display beside option_txt
+        """
+        self.prefix = prefix
+        self.setting = setting
+        self.submenu = submenu
+        self.option_gfx = gfx
+
+        self.is_writable = False
+
+    def __str__(self):
+        return f"Setting Chooser for {self.prefix}{self.setting}"
+
+    def set_editable(self, can_edit):
+        """Set whether or not we can write to this setting
+
+        @param can_edit  If True, we can write a new value
+        """
+        self.is_writable = can_edit
+
+    def is_editable(self):
+        return self.is_writable
+
+    def draw(self):
+        """Draw the menu to the screen
+
+        The OLED must be cleared before calling this function. You must call oled.show() after
+        calling this function
+        """
+
+        text_left = 0
+
+        oled.text(f"{self.prefix}{self.setting.display_name}", 0, 0)
+
+        if self.option_gfx is not None:
+            # draw the option thumbnail to the screen
+            text_left = 14
+            if self.is_writable:
+                img = k2.choice(self.option_gfx)
+            else:
+                key = self.setting.get_value()
+                img = self.option_gfx[key]
+            imgFB = FrameBuffer(img, 12, 12, MONO_HLSB)
+            oled.blit(imgFB, 0, SELECT_OPTION_Y)
+
+
+        if self.is_writable:
+            # draw the selection in inverted text
+            selected_item = k2.choice(self.setting.display_options)
+            choice_text = f"{selected_item}"
+            text_width = len(choice_text)*CHAR_WIDTH
+
+            oled.fill_rect(text_left, SELECT_OPTION_Y, text_left+text_width+3, CHAR_HEIGHT+4, 1)
+            oled.text(choice_text, text_left+1, SELECT_OPTION_Y+2, 0)
+        else:
+            # draw the selection in normal text
+            choice_text = f"{self.setting.get_display_value()}"
+            oled.text(choice_text, text_left+1, SELECT_OPTION_Y+2, 1)
+
+    def on_click(self):
+        if self.is_writable:
+            self.set_editable(False)
+            selected_index = k2.choice(list(range(len(self.setting))))
+            self.setting.choose(selected_index)
+        else:
+            self.set_editable(True)
```

Comment:
Yeah, that's a mild frustration I've had with the menu too.  It would be nice if there was an encoder instead of one of the knobs, but alas.

I'll look into the KnobBank class, but it likely won't get changed for a while.

Reply:
KnobBank is the better way to go, but one way I've addressed this is having a script "previous knob value" variable and use that to detect if the knob has moved before assigning a value to the current editable parameter. 

https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/polyrhythmic_sequencer.py#L278-L285

Also note that any time I change the page and therefore change the meaning of the previous value, I set `self._prev_k2` to None to ignore the first read on the first loop of the new page.

https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/polyrhythmic_sequencer.py#L167-L170

Reply:
I've just pushed a revision that uses the `KnobBank` class.  I'm not 100% sure I'm using it correctly; when jumping between menu levels the knob seems unresponsive for longer than it should (sometime I need to sweep it fully in both directions before it starts scrolling again). That might just be my own unfamiliarity with how it's intended to work though.

If one of you wants to pull the new version and let me know if the menu experience is better I'd appreciate it.

Reply:
I was suggesting using the `KnobBank` class for the individual settings controlled by K2. I think that it makes less sense for the menus and K1. I agree that an encoder would be ideal here.

Reply:
I'm actually finding the menu navigation easier with the bank on K1; when jumping up/down frequently it's nice to reliably be on the same setting you previously were instead of randomly skipping.

I'm afraid I don't really see the benefit to adding the knob bank to K2 though; if you've pressed the button to change the setting, I'd rather have the setting be freely changeable right away, rather than needing to scroll to unlock it before I can change the value.  And if you accidentally press the button you can long-press B2 to change levels & cancel the operation anyway.

Reply:
I'll admit that I have only run this script on my dev europi, which isn't in my rack, so I don't have the the full experience of what it is like to adjust settings in a creative context, rather than an 'explore the menu' context. So I'll rectify that.

I was imaging that when adjusting a setting, and it jumped to the current value of the knob that would be annoying in use. Especially if you don't remember the exact setting before. The cancel thing is nice, though not intuitive.

There are just not enough controls on a europi to make this kind of menu driven UI easy to use. Any decision you make will be full of compromise.

Reply:
Yeah, it's never going to be perfect. I think the trade-off of immediately being able to change a setting is probably more important to me than having the initial value per-setting stay consistent.  I don't mind the bank on K1 though; I've been trying it out for the last couple of days and I'm getting more used to it now.